### PR TITLE
[java] JUnitAssertionsShouldIncludeMessage - False positive with assertEquals and JUnit5

### LINKED ISSUE #1009: [java] JUnitAssertionsShouldIncludeMessage - False positive with assertEquals and JUnit5
**Affects PMD Version:** 6.1.0

**Rule:** JUnitAssertionsShouldIncludeMessage

**Additional version:**  
- JUnit : 5.1.0
- Gradle : 4.6

**Description:** 
With junit 5.1 the assertion message is on the last argument .
Here is the link to the javadoc : <https://junit.org/junit5/docs/5.0.0-M3/api/org/junit/jupiter/api/Assertions.html#assertEquals-java.lang.Object-java.lang.Object-java.lang.String->
In my sample, the rule should be ok, but when i check it, i have "_JUnit assertions should include a message_" exception.

**Result files:**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<pmd xmlns="http://pmd.sourceforge.net/report/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/report/2.0.0 http://pmd.sourceforge.net/report_2_0_0.xsd"
    version="6.1.0" timestamp="2018-03-27T18:48:51.560">
<file name="D:\Project\Java\issuePmd\src\test\java\SimpleTest.java">
<violation beginline="9" endline="9" begincolumn="9" endcolumn="20" rule="JUnitAssertionsShouldIncludeMessage" ruleset="Best Practices" package="test.java" class="SimpleTest" method="simpleMethod" externalInfoUrl="https://pmd.github.io/pmd-6.1.0/pmd_rules_java_bestpractices.html#junitassertionsshouldincludemessage" priority="3">
JUnit assertions should include a message
</violation>
</file>
</pmd>
```

**Code Sample demonstrating the issue:**
```java
package test.java;

import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.Test;

class SimpleTest {
    @Test
    public void simpleMethod() {
        assertEquals(0, 1, "Not equals 0 != 1");
    }
}
```

**Running PMD through:** *Gradle*

- build.gradle

```gradle
apply plugin: 'java'
apply plugin: 'eclipse'
apply plugin: 'pmd'

ext {
    junitJupiterVersion = '5.1.0'
}

repositories {
    mavenLocal()
    mavenCentral()
}

pmd{
    toolVersion = "6.1.0"
    ruleSetFiles = files("config/pmd/defaultRuleSet.xml")
    ruleSets = []
}

dependencies {
    // JUnit Jupiter API and TestEngine implementation
    testCompile("org.junit.jupiter:junit-jupiter-api:${junitJupiterVersion}")
    testRuntime("org.junit.jupiter:junit-jupiter-engine:${junitJupiterVersion}")
}

test {
    useJUnitPlatform()
}
```

- defaultRuleSet.xml
```xml
<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    name="Android Application Rules"
    xmlns="http://pmd.sf.net/ruleset/1.0.0"
    xsi:noNamespaceSchemaLocation="http://pmd.sf.net/ruleset_xml_schema.xsd"
    xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 
                        http://pmd.sf.net/ruleset_xml_schema.xsd" >

    <description>Standard rule set</description>
    <rule ref="category/java/bestpractices.xml" />
</ruleset>
```

### PR DESCRIPTION
**Pull request by djydewang**
Resolves #1009

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitAssertionsShouldIncludeMessageRule.java [hunk 1]
```diff
@@ -65,8 +65,11 @@ public JUnitAssertionsShouldIncludeMessageRule() {
         checks.add(new AssertionCall("assertEquals", 3) {
             @Override
             protected boolean isException(ASTArguments node) {
-                ASTExpression firstArgument = node.getFirstDescendantOfType(ASTExpression.class);
-                return firstArgument.getType() == null || firstArgument.getType() == String.class;
+                List<ASTExpression> arguments = node.findDescendantsOfType(ASTExpression.class);
+                boolean isExceptionJunit4 = isStringTypeOrNull(arguments.get(0));
+                boolean isExceptionJunit5 = isStringTypeOrNull(arguments.get(2));
+
+                return isExceptionJunit4 || isExceptionJunit5;
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitAssertionsShouldIncludeMessageRule.java]: That's okay, but keep in mind, that it's actually not 100% correct:
We would need to check either for the first argument (junit4) or last argument (junit5) to be a string - after we determined the Junit version. But the combination "first argument is a String" and "JUnit5" would be wrong (and would not even compile).
However, I think, we can live with that simplification: We know here already, that we have in total 3 arguments. And the compiler will ensure, that the correct overload is used (where ever the String is).

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml [hunk 2]
```diff
@@ -426,4 +426,18 @@ public class TestNgTestCase {
 }
         ]]></code>
     </test-code>
+    <test-code>
+        <description><![CDATA[
+Issue #1009: JUnitAssertionsShouldIncludeMessage - False positive with assertEquals
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml]: I'll change the title of the issue to reflect, that it's actually a false positive involving JUnit5. Could you please change the description here as well?
Btw. the CDATA is not needed here - without it, the description becomes a bit more readable.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml [hunk 3]
```diff
@@ -426,4 +426,18 @@ public class TestNgTestCase {
...
+     ]]></description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import junit.framework.TestCase;
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml]: I'd prefer to have here a actual JUnit5 example - the TestCase class is actually Junit3.

Just use the code snipped from the issue:

```java
package test.java;

import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.Test;

class SimpleTest {
    @Test
    public void simpleMethod() {
        assertEquals(0, 1, "Not equals 0 != 1");
    }
}
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml (hunk 2 above)
Comment (djydewang) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml]: Done.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml (hunk 3 above)
Comment (djydewang) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml]: Done.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml (hunk 2 above)
Comment (djydewang) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnitAssertionsShouldIncludeMessage.xml]: Done.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitAssertionsShouldIncludeMessageRule.java (hunk 1 above)
Comment (djydewang) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitAssertionsShouldIncludeMessageRule.java]: > However, I think, we can live with that simplification: We know here already, that we have in total 3 arguments. And the compiler will ensure, that the correct overload is used (where ever the String is).

That's what I think, PMD checks should be based on programs that can be compiled successfully.

**Comment by adangel** [comment]
Merged! Thanks again for your contribution!

**Pull request merged**